types/src/transaction/mod.rs

```diff
@@ -887,34 +884,31 @@ impl From<TransactionToCommit> for crate::proto::types::TransactionToCommit {
 /// 3. The list has 2+ transactions/transaction_infos. The both proofs must exist.
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct TransactionListWithProof {
-    pub transaction_and_infos: Vec<(Transaction, TransactionInfo)>,
+    pub transactions: Vec<Transaction>,
     pub events: Option<Vec<Vec<ContractEvent>>>,
     pub first_transaction_version: Option<Version>,
-    pub proof_of_first_transaction: Option<TransactionAccumulatorProof>,
-    pub proof_of_last_transaction: Option<TransactionAccumulatorProof>,
+    pub proof: TransactionListProof,
 }
 
 impl TransactionListWithProof {
     /// Constructor.
     pub fn new(
-        transaction_and_infos: Vec<(Transaction, TransactionInfo)>,
+        transactions: Vec<Transaction>,
         events: Option<Vec<Vec<ContractEvent>>>,
         first_transaction_version: Option<Version>,
-        proof_of_first_transaction: Option<TransactionAccumulatorProof>,
-        proof_of_last_transaction: Option<TransactionAccumulatorProof>,
+        proof: TransactionListProof,
     ) -> Self {
         Self {
-            transaction_and_infos,
+            transactions,
             events,
             first_transaction_version,
-            proof_of_first_transaction,
-            proof_of_last_transaction,
+            proof,
```

Comment:
I think it'd better check the length of `transactions`, `events` and `transaction_infos` here. So most `ensure`s are not necessary for `verify()`.